frontend/data.ts

```diff
@@ -1,101 +1,99 @@
+        await putShape(writeStorage(tx), args);
       }
-    );
+    ),
 
-    this.moveShape = rep.register(
+    moveShape: rep.register(
       "moveShape",
-      async (tx: WriteTransaction, args: MoveShapeArgs) => {
-        await moveShape(this.writeStorage(tx), args);
+      async (
+        tx: WriteTransaction,
+        args: { id: string; dx: number; dy: number }
+      ) => {
+        await moveShape(writeStorage(tx), args);
       }
-    );
+    ),
 
-    this.overShape = rep.register(
+    overShape: rep.register(
       "overShape",
-      async (tx: WriteTransaction, args: OverShapeArgs) => {
-        await overShape(this.writeStorage(tx), args);
+      async (
+        tx: WriteTransaction,
+        args: { clientID: string; shapeID: string }
+      ) => {
+        await overShape(writeStorage(tx), args);
       }
-    );
-  }
+    ),
 
-  // TODO: Is there a way for Typescript to infer this from the assignment in constructor?
-  readonly createShape: (args: CreateShapeArgs) => Promise<void>;
-  readonly moveShape: (args: MoveShapeArgs) => Promise<void>;
-  readonly overShape: (args: OverShapeArgs) => Promise<void>;
+    // subscriptions
+    useShapeIDs(): Array<string> {
+      return useSubscribe(
+        rep,
+        async (tx: ReadTransaction) => {
+          const shapes = await tx.scanAll({ prefix: "shape-" });
+          return shapes.map(([k, _]) => k.split("-")[1]);
+        },
+        []
+      );
+    },
 
-  useShapeIDs(): Array<string> {
-    return useSubscribe(
-      this.rep,
-      async (tx: ReadTransaction) => {
-        const shapes = await tx.scanAll({ prefix: "shape-" });
-        return shapes.map(([k, _]) => k.split("-")[1]);
-      },
-      []
-    );
-  }
-
-  useShapeByID(id: string): Shape | null {
-    return useSubscribe(
-      this.rep,
-      (tx: ReadTransaction) => {
-        return getShape(this.readStorage(tx), id);
-      },
-      null
-    );
-  }
+    useShapeByID(id: string): Shape | null {
+      return useSubscribe(
+        rep,
+        (tx: ReadTransaction) => {
+          return getShape(readStorage(tx), id);
+        },
+        null
+      );
+    },
 
-  useOverShapeID(): string | null {
-    return useSubscribe(this.rep, async (tx: ReadTransaction) => {
-      return (await getClientState(this.readStorage(tx), this.clientID)).overID;
-    });
-  }
+    useOverShapeID(): string | null {
+      return useSubscribe(
+        rep,
+        async (tx: ReadTransaction) => {
+          return (await getClientState(readStorage(tx), clientID)).overID;
+        },
+        null
+      );
+    },
+  };
+}
 
-  private readStorage(tx: ReadTransaction): Storage {
-    return {
-      getObject: tx.get.bind(tx),
-      putObject: () => {
-        throw new Error("Cannot write inside ReadTransaction");
-      },
-    };
-  }
+function readStorage(tx: ReadTransaction): Storage {
+  return {
+    getObject: tx.get.bind(tx),
```

Comment:
bind is slower but 🤷 

I don't think I ever felt the need to use bind once I got used to arrow functions.

Reply:
I love bind so much, but if it's slower, I'll go with your way.